mantidproject/mantid
# Q summation cropping issue

This PR adds an option to `ReflectometryReductionOneAuto` and `ReflectometryReductionOne` to control whether partial bins at the ends of the output array should be included or not when summing in Q. Previously they were always included with no option to exclude them. They are now excluded by default and can be included by enabling `IncludePartialBins`.

We use the appropriate range from the figure in the linked issue, that is, (lambda_f1, lambda_f2) to include partial bins or (lambda_c1, lambda_c2) to exclude them.

**Report to:** Jos at ISIS

**To test:**

Run the script below and check you get the expected plots. Also check that the `ISIS Reflectometry` GUI passes the correct setting through when you tick or untick `IncludePartialBins` on the Settings tab.

```
import matplotlib.pyplot as plt
inst='OFFSPEC'
LoadISISNexus(Filename='OFFSPEC46911', OutputWorkspace='TRANS_46911')


def plotResults(title, ws_list):
    fig, axes = plt.subplots(subplot_kw={'projection': 'mantid'})
    axes.set_xscale('log')
    axes.set_yscale('log')
    axes.set_xlim(0.005,0.1)
    axes.set_ylim(0.00001,1)
    axes.set_title(title)
    for ws in ws_list:
        axes.plot(mtd[ws])
    fig.show()

def runReduction(run, theta, suffix='',partial=0):
    name=str(run)+suffix
    LoadISISNexus(Filename=inst+str(run), OutputWorkspace='TOF_'+name)
    ReflectometryReductionOneAuto(InputWorkspace='TOF_'+name, SummationType='SumInQ', ReductionType='DivergentBeam', ProcessingInstructions='390-415',
        ThetaIn=theta, WavelengthMin=2, WavelengthMax=14, I0MonitorIndex=1, 
        MonitorBackgroundWavelengthMin=15, MonitorBackgroundWavelengthMax=20, MonitorIntegrationWavelengthMin=2, MonitorIntegrationWavelengthMax=14, 
        FirstTransmissionRun='TRANS_46911', StartOverlap=10, EndOverlap=12,
        MomentumTransferStep=0.029999999999999999, 
        ScaleFactor=1, 
        OutputWorkspaceBinned='IvsQ_binned_'+name, OutputWorkspace='IvsQ_'+name, OutputWorkspaceWavelength='IvsLam_'+name,
        IncludePartialBins=partial
        )

suffix=''
runReduction(46895, 0.3,suffix, 1)
runReduction(46896,0.9,suffix, 1)
runReduction(46897, 2.3,suffix, 1)
Stitch1DMany(InputWorkspaces='IvsQ_46895,IvsQ_46896,IvsQ_46897', OutputWorkspace='IvsQ_46895_46896_46897', Params='-0.03', OutScaleFactors='0.162718,0.0260704')
plotResults('Include full input range', ['IvsQ_binned_46895','IvsQ_binned_46896','IvsQ_binned_46897','IvsQ_46895_46896_46897'])

suffix='_cropped'
runReduction(46895, 0.3,suffix)
runReduction(46896,0.9,suffix)
runReduction(46897, 2.3,suffix)
Stitch1DMany(InputWorkspaces='IvsQ_46895_cropped,IvsQ_46896_cropped,IvsQ_46897_cropped', OutputWorkspace='IvsQ_46895_46896_46897_cropped', Params='-0.03', OutScaleFactors='0.162718,0.0260704')
plotResults('Exclude partial bins', ['IvsQ_binned_46895_cropped','IvsQ_binned_46896_cropped','IvsQ_binned_46897_cropped','IvsQ_46895_46896_46897_cropped'])
```

This should produce the following two outputs with partial bins excluded/included:
![q_sum_cropping_error_fixed](https://user-images.githubusercontent.com/12895056/41662931-214edce2-749a-11e8-9cfd-964b92b84e4b.png)
![sum_in_q_cropping_issue](https://user-images.githubusercontent.com/12895056/41662934-2380e406-749a-11e8-9520-375621377ef8.png)


Fixes #22541

Does this update require release notes?


---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####


##### Functional Tests #####


Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
